components/concept/fetch/src/main/java/mozilla/components/concept/fetch/Request.kt

```
@@ -0,0 +1,111 @@
+ * infinite timeout.
+ * @property readTimeout A timeout to be used when reading from the resource. If the timeout expires before there is
+ * data available for read, a java.net.SocketTimeoutException is raised. A timeout of zero is interpreted as an infinite
+ * timeout.
+ * @property body An optional body to be send with the request.
+ * @property redirect Whether the [Client] should follow redirects (HTTP 3xx) for this request or not.
+ */
+data class Request(
+    val url: String,
+    val method: Method = Method.GET,
+    val headers: MutableHeaders? = MutableHeaders(),
+    val connectTimeout: Pair<Long, TimeUnit>? = null,
+    val readTimeout: Pair<Long, TimeUnit>? = null,
+    val body: Body? = null,
+    val redirect: Redirect = Redirect.FOLLOW
+) {
+    /**
+     * A [Body] to be send with the [Request].
+     *
+     * @param stream A stream that will be read and send to the resource.
+     */
+    class Body(
+        private val stream: InputStream
+    ) : Closeable {
+        companion object {
+            /**
+             * Create a [Body] from the provided [String].
+             */
+            fun fromString(value: String): Body = Body(value.byteInputStream())
+
+            /**
+             * Create a [Body] from the provided [File].
+             */
+            fun fromFile(file: File): Body = Body(file.inputStream())
+        }
+
+        /**
+         * Executes the given [block] function on the body's stream and then closes it down correctly whether an
+         * exception is thrown or not.
+         */
+        fun <R> useStream(block: (InputStream) -> R): R = use {
+            block(stream)
+        }
+
+        /**
+         * Closes this body and releases any system resources associated with it.
+         */
+        override fun close() {
+            try {
+                stream.close()
+            } catch (e: IOException) {
+                // Ignore
```

Is it not better to propagate the exception?
It could hide unexpected behavior for the client, like failing silently.

---

In all other cases I do that. For `close` I'm not sure: We have read and send everything. There's no error with the request and there's not really anything the calling code can do about that exception.

---

👍